src/Servers/test/FunctionalTests/HelloWorldTest.cs

```diff
@@ -108,4 +109,70 @@ public async Task HelloWorld(TestVariant variant)
             }
         }
     }
+
+    public static TestMatrix SelfHostTestVariants
+        => TestMatrix.ForServers(ServerType.Kestrel, ServerType.HttpSys)
+            .WithTfms(Tfm.Default)
+            .WithApplicationTypes(ApplicationType.Portable)
+            .WithAllHostingModels()
+            .WithAllArchitectures();
+
+    [ConditionalTheory]
+    [MemberData(nameof(SelfHostTestVariants))]
+    public async Task ApplicationException(TestVariant variant)
+    {
+        var testName = $"ApplicationException_{variant.Server}_{variant.Tfm}_{variant.Architecture}_{variant.ApplicationType}";
+        using (StartLog(out var loggerFactory,
+            variant.Server == ServerType.Nginx ? LogLevel.Trace : LogLevel.Debug, // https://github.com/aspnet/ServerTests/issues/144
+            testName))
+        {
+            var logger = loggerFactory.CreateLogger("ApplicationException");
+
+            var deploymentParameters = new DeploymentParameters(variant)
+            {
+                ApplicationPath = Helpers.GetApplicationPath()
+            };
+
+            if (variant.Server == ServerType.Nginx)
+            {
+                deploymentParameters.ServerConfigTemplateContent = Helpers.GetNginxConfigContent("nginx.conf");
+            }
+
+            if (variant.Server == ServerType.IISExpress)
+            {
+                deploymentParameters.EnvironmentVariables[DebugEnvironmentVariable] = "console";
+            }
+
+            using (var deployer = new SelfHostDeployer(deploymentParameters, loggerFactory))
+            {
+                var output = string.Empty;
+                deployer.ProcessOutputListener = (data) =>
+                {
+                    if (!string.IsNullOrWhiteSpace(data))
+                    {
+                        output += data + '\n';
+                    }
+                };
+
+                var deploymentResult = await deployer.DeployAsync();
+
+                // Request to base address and check if various parts of the body are rendered & measure the cold startup time.
+                using (var response = await RetryHelper.RetryRequest(() =>
+                {
+                    return deploymentResult.HttpClient.GetAsync("/throwexception");
+                }, logger, deploymentResult.HostShutdownToken))
+                {
+                    Assert.Equal(HttpStatusCode.InternalServerError, response.StatusCode);
+
+                    var body = await response.Content.ReadAsStringAsync();
+                    Assert.Empty(body);
+                }
+
+                await Task.Delay(50); // wait a bit for the server to finish the request
```

Comment:
We try to avoid doing arbitrary delays like this. Slower machines or machines running lots of tests at once could need longer than 50ms.

Would shutting down the server and waiting for that to complete be a better alternative?

Reply:
Thanks for the suggestion, yes that works.